index.ts

```diff
@@ -0,0 +1,476 @@
+type MoveSource = string;
+
+interface EventInfo {
+	generation: Generation;
+	level?: number;
+	shiny?: boolean | 1;
+	gender?: GenderName;
+	nature?: NatureName;
+	ivs?: Partial<StatsTable>;
+	perfectIVs?: number;
+	isHidden?: boolean;
+	abilities?: ID[];
+	maxEggMoves?: number;
+	moves?: ID[];
+	pokeball?: ID;
+	from?: string;
+}
+
+interface Species extends Data {
+	type 'Species';
+
+	abilities: SpeciesAbilities;
+	addedType?: string;
+	baseForme?: string;
+	baseSpecies?: string;
+	baseStats: StatsTable;
+	battleOnly?: boolean
+	canHatch?: boolean;
+	color: string;
+	cosmeticForms?: ID[];
+	doublesTier?: TierName;
+	eggGroups: string[];
+	encounters?: EventInfo[]
+	eventOnly?: boolean
+	eventPokemon?: EventInfo[];
+	evoCondition?: string;
+	evoItem?: string;
+	evoLevel?: number;
+	evoMove?: string;
+	evos?: ID[];
+	evoType?:
+	formeid: string;
+	formeLetter?: string;
+	forme?: string;
+	gender?: GenderName;
+	genderRatio?: {M: number, F: number};
+	heightm: number;
+	isMega?: boolean;
+	isPrimal?: boolean;
+	isTotem: boolean;
+	learnset?: {[moveid: string]: MoveSource[]};
+	maleOnlyHidden?: boolean
+	maxHP?: number;
+	nfe: boolean;
+	otherFormes?: ID[];
+	prevo?: ID;
+	requiredAbility?: ID;
+	requiredItems?: ID[];
+	requiredMove?: ID;
+	tier?: TierName;
+	types: TypeName[];
+	unreleasedHidden?: boolean
+	weightkg: number;
+}
+
+// TODO: name collision :(
+class Species {
+	static forGen(gen: Generation): Promise<Species>;
+
+	getName(name: string): string;
+	get(name: string): Species | undefined;
+	getByID(id: ID): Species | undefined;
+	next(): Species | true;
+}
+
+// stats.ts
+
+type StatName = 'hp' | 'atk' | 'def' | 'spa' | 'spd' | 'spe';
+type StatsTable<T = number> = {[stat in StatName]: T};
+type BoostName = Exclude<StatName, 'hp'> | 'accuracy' | 'evasion';
+type BoostsTable<T = number> = {[boost in BoostName]: T};
+
+export class Stats {
+	static calc(stat: Stat, base: number, iv: number, ev: number, level: number, gen?: Generation);
+	static calc(stat: Stat, base: number, iv: number, ev: number, level: number, nature: Nature, gen?: Generation);
+
+	static get(name: string): StatName | undefined;
+	static display(s: string, full: boolean = true): string;
+
+  static fill(p: Partial<StatsTable<T>>, val: T): StatsTable<T>;
+	static itod(iv: number): number;
+	static dtoi(dv: number): number;
+	static getHPDV(pivs: Partial<StatsTable>): number;
+}
+
+// statuses.ts
+
+type StatusName = 'par' | 'psn' | 'frz' | 'slp' | 'brn' | 'tox';
+
+// sets.ts
+
+// TODO: other than species/moves the rest can be optional?
+interface PokemonSet<T extends string = string> {
+	name: string;
+	species: T;
+	item: T;
+	ability: T;
+	moves: T[];
+	nature: T;
+	gender: GenderName;
+	evs: StatsTable;
+	ivs: StatsTable;
+	level: number;
+	shiny?: boolean;
+	happiness?: number;
+	hpType?: string;
+	pokeball?: T;
+}
+
+class Sets {
+	static forGen(gen: Generation): Promise<Teams>;
+
+	packSet(s: PokemonSet<ID  | string>): string;
```

Comment:
what is packing/unpacking, and how does it differ from exporting/importing?

Reply:
packing/unpacking is machine readable vs. import/export which is human readable. import/export is what you see on pokepastes, packing is what is sent internally over the wire: see https://github.com/Zarel/Pokemon-Showdown/blob/master/PROTOCOL.md#team-format for an example. i believe the packed format cuts down on 90% of the size

Reply:
i see. well, this is just a stylistic nit, but maybe just have `exportSet` and `importSet` and have them take an `ExportType` which could be `compact` or `humanReadable` (idk)